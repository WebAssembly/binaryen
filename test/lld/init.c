volatile int x, y;

__attribute__((constructor))
void init_x() {
    x = 14;
}

__attribute__((constructor))
void init_y() {
    y = 144;
}

int main() {
    return x + y;
}
