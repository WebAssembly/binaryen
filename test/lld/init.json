{ "asmConsts": {},"staticBump": 8, "initializers": ["init_x", "init_y"] }
