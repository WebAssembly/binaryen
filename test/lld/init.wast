(module
 (type $0 (func))
 (type $1 (func (result i32)))
 (global $global$0 (mut i32) (i32.const 66576))
 (global $global$1 i32 (i32.const 66576))
 (global $global$2 i32 (i32.const 1032))
 (table 1 1 anyfunc)
 (memory $0 2)
 (data (i32.const 1024) "\00\00\00\00\00\00\00\00")
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (func $init_x (type $0)
  (i32.store offset=1024
   (i32.const 0)
   (i32.const 14)
  )
 )
 (func $init_y (type $0)
  (i32.store offset=1028
   (i32.const 0)
   (i32.const 144)
  )
 )
 (func $main (type $1) (result i32)
  (i32.add
   (i32.load offset=1024
    (i32.const 0)
   )
   (i32.load offset=1028
    (i32.const 0)
   )
  )
 )
 (func $__wasm_call_ctors (type $0)
  (call $init_x)
  (call $init_y)
 )
)

