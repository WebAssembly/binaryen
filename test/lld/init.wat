(module
 (type $none_=>_none (func))
 (type $none_=>_i32 (func (result i32)))
 (type $i32_i32_=>_i32 (func (param i32 i32) (result i32)))
 (memory $0 2)
 (table $0 1 1 funcref)
 (global $global$0 (mut i32) (i32.const 66112))
 (global $global$1 i32 (i32.const 576))
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__data_end" (global $global$1))
 (func $__wasm_call_ctors
  (call $init_x)
  (call $init_y)
 )
 (func $init_x
  (i32.store offset=568
   (i32.const 0)
   (i32.const 14)
  )
 )
 (func $init_y
  (i32.store offset=572
   (i32.const 0)
   (i32.const 144)
  )
 )
 (func $__original_main (result i32)
  (i32.add
   (i32.load offset=568
    (i32.const 0)
   )
   (i32.load offset=572
    (i32.const 0)
   )
  )
 )
 (func $main (param $0 i32) (param $1 i32) (result i32)
  (call $__original_main)
 )
 ;; custom section "producers", size 112
)

