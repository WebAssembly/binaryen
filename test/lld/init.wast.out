(module
 (type $0 (func))
 (type $1 (func (result i32)))
 (type $2 (func (param i32 i32) (result i32)))
 (memory $0 2)
 (data (i32.const 568) "\00\00\00\00\00\00\00\00")
 (table $0 1 1 funcref)
 (global $global$0 (mut i32) (i32.const 16384))
 (global $global$1 i32 (i32.const 66112))
 (global $global$2 i32 (i32.const 576))
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (export "main" (func $main))
 (export "stackSave" (func $stackSave))
 (export "stackAlloc" (func $stackAlloc))
 (export "stackRestore" (func $stackRestore))
 (export "__growWasmMemory" (func $__growWasmMemory))
 (func $__wasm_call_ctors (; 0 ;) (type $0)
  (call $init_x)
  (call $init_y)
 )
 (func $init_x (; 1 ;) (type $0)
  (i32.store offset=568
   (i32.const 0)
   (i32.const 14)
  )
 )
 (func $init_y (; 2 ;) (type $0)
  (i32.store offset=572
   (i32.const 0)
   (i32.const 144)
  )
 )
 (func $__original_main (; 3 ;) (type $1) (result i32)
  (i32.add
   (i32.load offset=568
    (i32.const 0)
   )
   (i32.load offset=572
    (i32.const 0)
   )
  )
 )
 (func $main (; 4 ;) (type $2) (param $0 i32) (param $1 i32) (result i32)
  (call $__original_main)
 )
 (func $stackSave (; 5 ;) (result i32)
  (global.get $global$0)
 )
 (func $stackAlloc (; 6 ;) (param $0 i32) (result i32)
  (local $1 i32)
  (global.set $global$0
   (local.tee $1
    (i32.and
     (i32.sub
      (global.get $global$0)
      (local.get $0)
     )
     (i32.const -16)
    )
   )
  )
  (local.get $1)
 )
 (func $stackRestore (; 7 ;) (param $0 i32)
  (global.set $global$0
   (local.get $0)
  )
 )
 (func $__growWasmMemory (; 8 ;) (param $newSize i32) (result i32)
  (grow_memory
   (local.get $newSize)
  )
 )
)
(;
--BEGIN METADATA --
{
  "staticBump": 8,
  "tableSize": 1,
  "initializers": [
    "__wasm_call_ctors"
  ],
  "declares": [
  ],
  "externs": [
  ],
  "implementedFunctions": [
    "___wasm_call_ctors",
    "_main",
    "_stackSave",
    "_stackAlloc",
    "_stackRestore",
    "___growWasmMemory"
  ],
  "exports": [
    "memory",
    "__wasm_call_ctors",
    "__heap_base",
    "__data_end",
    "main",
    "stackSave",
    "stackAlloc",
    "stackRestore",
    "__growWasmMemory"
  ],
  "invokeFuncs": [
  ]
}
-- END METADATA --
;)
